File: Src/Engine/Editor/GUI/Docking/DockHintWindow.cpp


#include "DockHintWindow.h"

#include "DockWindow.h"
#include "FloatWindowDockPanel.h"
#include "MasterDockPanel.h"
#include "Runtime/Graphics/GraphicWindow.h"
#include "Runtime/UI/GUI/Style.h"
#include "Runtime/UI/GUI/WindowRootControl.h"

namespace SE::Editor
{
	DockHintWindow::DockHintWindow(FloatWindowDockPanel* toMove)
	{
		_OnMouseUpCall = CreateFunc<DockHintWindow, &DockHintWindow::OnMouseUp>(this);
		_OnMouseMoveCall = CreateFunc<DockHintWindow, &DockHintWindow::OnMouseMove>(this);
		_OnLostFocusCall = CreateFunc<DockHintWindow, &DockHintWindow::OnLostFocus>(this);

		_toMove = toMove;
		_toSet = DockState::Float;
		GraphicWindow* window = toMove->Window->Window();

		// Remove focus from drag target
		_toMove->Focus();
		_toMove->Defocus();

		// Focus window
		window->Focus();

		// Check if window is maximized and restore window.
		if (window->IsMaximized())
		{
			// Restore window and set position to mouse.
			Float2 mousePos = window->GetMousePosition();
			Float2 previousSize = window->GetSize();
			window->Restore();
			window->SetPosition(Platform::GetMousePosition() - mousePos * window->GetSize() / previousSize);
		}

		// Calculate dragging offset and move window to the destination position
		Float2 mouseScreenPosition = Platform::GetMousePosition();

		// If the _toMove window was not focused when initializing this window, the result vector only contains zeros
		// and to prevent a failure, we need to perform an update for the drag offset at later time which will be done in the OnMouseMove event handler.
		if (mouseScreenPosition != Float2::Zero)
			CalculateDragOffset(mouseScreenPosition);
		else
			_lateDragOffsetUpdate = true;

		// Get initial size
		_defaultWindowSize = window->GetSize();

		// Init proxy window
		Proxy::Init(_defaultWindowSize);

		// Bind events
		Proxy::Window->MouseUpEvent.BindUnique(_OnMouseUpCall);;
		Proxy::Window->MouseMoveEvent.BindUnique(_OnMouseMoveCall);
		Proxy::Window->LostFocusEvent.BindUnique(_OnLostFocusCall);

		// Start tracking mouse
		Proxy::Window->StartTrackingMouse(false);

		// Update window GUI
		Proxy::Window->GetGUI()->PerformLayout();

		// Update rectangles
		UpdateRects();

		// Hide base window
		window->Hide();

		// Enable hit window presentation
		Proxy::Window->SetRenderingEnabled(true);
		Proxy::Window->Show();
		Proxy::Window->Focus();
	}
	
	void DockHintWindow::CalculateDragOffset(Float2 mouseScreenPosition)
	{
		Float2 baseWinPos = _toMove->Window->Window()->GetPosition();
		_dragOffset = mouseScreenPosition - baseWinPos;
	}
	
	void DockHintWindow::UpdateRects()
    {
        // Cache mouse position
        _mouse = Platform::GetMousePosition();

        // Check intersection with any dock panel
        Float2 uiMouse = _mouse;
        _toDock = _toMove->MasterPanel->HitTest(uiMouse, _toMove);

        // Check dock state to use
        bool showProxyHints = _toDock != nullptr;
        bool showBorderHints = showProxyHints;
        bool showCenterHint = showProxyHints;
        if (showProxyHints)
        {
            // If moved window has not only tabs but also child panels disable docking as tab
            if (_toMove->ChildPanelsCount > 0)
                showCenterHint = false;

            // Disable docking windows with one or more dock panels inside
            if (_toMove->ChildPanelsCount > 0)
                showBorderHints = false;

            // Get dock area
            _rectDock = _toDock->DockAreaBounds;

            // Cache dock rectangles
            Float2 size = _rectDock.Size;
            Float2 offset = _rectDock.Location;
            float borderMargin = 4.0f;
            float hintWindowsSize = Proxy::HintWindowsSize * Platform::GetDpiScale();
            float hintWindowsSize2 = hintWindowsSize * 0.5f;
            float centerX = size.x * 0.5f;
            float centerY = size.y * 0.5f;
            _rUpper = Rectangle(centerX - hintWindowsSize2, borderMargin, hintWindowsSize, hintWindowsSize) + offset;
            _rBottom = Rectangle(centerX - hintWindowsSize2, size.y - hintWindowsSize - borderMargin, hintWindowsSize, hintWindowsSize) + offset;
            _rLeft = Rectangle(borderMargin, centerY - hintWindowsSize2, hintWindowsSize, hintWindowsSize) + offset;
            _rRight = Rectangle(size.x - hintWindowsSize - borderMargin, centerY - hintWindowsSize2, hintWindowsSize, hintWindowsSize) + offset;
            _rCenter = Rectangle(centerX - hintWindowsSize2, centerY - hintWindowsSize2, hintWindowsSize, hintWindowsSize) + offset;

            // Hit test
            DockState toSet = DockState::Float;
            if (showBorderHints)
            {
                if (_rUpper.Contains(_mouse))
                    toSet = DockState::DockTop;
                else if (_rBottom.Contains(_mouse))
                    toSet = DockState::DockBottom;
                else if (_rLeft.Contains(_mouse))
                    toSet = DockState::DockLeft;
                else if (_rRight.Contains(_mouse))
                    toSet = DockState::DockRight;
            }
            if (showCenterHint && _rCenter.Contains(_mouse))
                toSet = DockState::DockFill;
            _toSet = toSet;

            // Show proxy hint windows
            Proxy::Down->SetPosition(_rBottom.Location);
            Proxy::Left->SetPosition(_rLeft.Location);
            Proxy::Right->SetPosition(_rRight.Location);
            Proxy::Up->SetPosition(_rUpper.Location);
            Proxy::Center->SetPosition(_rCenter.Location);
        }
        else
        {
            _toSet = DockState::Float;
        }

        // Update proxy hint windows visibility
        Proxy::Down->SetIsVisible(showProxyHints & showBorderHints);
        Proxy::Left->SetIsVisible(showProxyHints & showBorderHints);
        Proxy::Right->SetIsVisible(showProxyHints & showBorderHints);
        Proxy::Up->SetIsVisible(showProxyHints & showBorderHints);
        Proxy::Center->SetIsVisible(showProxyHints & showCenterHint);

        // Calculate proxy/dock/window rectangles
        if (_toDock == nullptr)
        {
            // Floating window over nothing
            _rectWindow = Rectangle(_mouse - _dragOffset, _defaultWindowSize);
        }
        else
        {
            if (_toSet == DockState::Float)
            {
                // Floating window over dock panel
                _rectWindow = Rectangle(_mouse - _dragOffset, _defaultWindowSize);
            }
            else
            {
                // Use only part of the dock panel to show hint
                _rectWindow = CalculateDockRect(_toSet, _rectDock);
            }
        }

        // Update proxy window
        Proxy::Window->SetClientBounds(_rectWindow);
    }
	
	void DockHintWindow::OnMouseUp(const Float2& location, MouseButton button)
	{
		if (button == MouseButton::Left)
		{
			Dispose();
		}
	}
	
	void DockHintWindow::OnMouseMove(const Float2& mousePos)
	{
		// Recalculate the drag offset because the current mouse screen position was invalid when we initialized the window
		if (_lateDragOffsetUpdate)
		{
			// Calculate dragging offset and move window to the destination position
			CalculateDragOffset(mousePos);

			// Reset state
			_lateDragOffsetUpdate = false;
		}

		UpdateRects();
	}
	
	void DockHintWindow::OnLostFocus()
	{
		Dispose();
		Delete(this);
	}
	
	void DockHintWindow::Dispose()
    {
        // End tracking mouse
        Proxy::Window->EndTrackingMouse();

        // Disable rendering
        Proxy::Window->SetRenderingEnabled(false);

        // Unbind events
        Proxy::Window->MouseUpEvent.Unbind(_OnMouseUpCall);
        Proxy::Window->MouseMoveEvent.Unbind(_OnMouseMoveCall);
        Proxy::Window->LostFocusEvent.Unbind(_OnLostFocusCall);

		_OnMouseMoveCall.Unbind();
		_OnMouseMoveCall.Unbind();
		_OnLostFocusCall.Unbind();

        // Hide the proxy
        Proxy::Hide();

        if (_toMove == nullptr)
            return;

        // Check if window won't be docked
        if (_toSet == DockState::Float)
        {
        	if (_toMove->Window == nullptr)
        	{
        		return;
        	}
            GraphicWindow* window = _toMove->Window->Window();
            if (window == nullptr)
                return;
            Float2 mouse = Platform::GetMousePosition();

            // Move base window
            window->SetPosition(mouse - _dragOffset);

            // Show base window
            window->Show();
        }
        else
        {
            bool hasNoChildPanels = _toMove->ChildPanelsCount == 0;

            // Check if window has only single tab
            if (hasNoChildPanels && _toMove->TabsCount == 1)
            {
                // Dock window
                _toMove->GetTab(0)->Show(_toSet, _toDock);
            }
            // Check if dock as tab and has no child panels
            else if (hasNoChildPanels && _toSet == DockState::DockFill)
            {
                // Dock all tabs
                while (_toMove->TabsCount > 0)
                {
                    _toMove->GetTab(0)->Show(DockState::DockFill, _toDock);
                }
            }
            else
            {
                DockWindow* selectedTab = _toMove->SelectedTab;

                // Dock the first tab into the target location
                DockWindow* firstTab = _toMove->GetTab(0);
                firstTab->Show(_toSet, _toDock);

                // Dock rest of the tabs
                while (_toMove->TabsCount > 0)
                {
                    _toMove->GetTab(0)->Show(DockState::DockFill, firstTab);
                }

                // Keep selected tab being selected
            	if (selectedTab != nullptr)
            	{
            		selectedTab->SelectTab();
            	}
            }

            // Focus target window
            _toDock->Root->Focus();
        }

        _toMove = nullptr;
    }
	
	DockHintWindow* DockHintWindow::Create(FloatWindowDockPanel* toMove)
	{
		ENGINE_ASSERT(toMove != nullptr);
		
		return New<DockHintWindow>(toMove);
	}
	
	DockHintWindow* DockHintWindow::Create(DockWindow* toMove)
	{
		ENGINE_ASSERT(toMove != nullptr);

		// Show floating
		toMove->ShowFloating();

		// Move window to the mouse position (with some offset for caption bar)
		WindowRootControl* window = static_cast<WindowRootControl*>(toMove->Root.operator->());
		Float2 mouse = Platform::GetMousePosition();
		window->Window()->SetPosition(mouse - Float2(8, 8));

		// Get floating panel
		FloatWindowDockPanel* floatingPanelToMove = TypeTryCast<FloatWindowDockPanel>(window->GetChild(0));

		return New<DockHintWindow>(floatingPanelToMove);
	}
	
	Rectangle DockHintWindow::CalculateDockRect(DockState state, Rectangle& rect)
	{
		Rectangle result = rect;
		switch (state)
		{
		case DockState::DockFill:
			result.Location.y += DockPanel::DefaultHeaderHeight;
			result.Size.y -= DockPanel::DefaultHeaderHeight;
			break;
		case DockState::DockTop:
			result.Size.y *= DockPanel::DefaultSplitterValue;
			break;
		case DockState::DockLeft:
			result.Size.x *= DockPanel::DefaultSplitterValue;
			break;
		case DockState::DockBottom:
			result.Location.y += result.Size.y * (1 - DockPanel::DefaultSplitterValue);
			result.Size.y *= DockPanel::DefaultSplitterValue;
			break;
		case DockState::DockRight:
			result.Location.x += result.Size.x * (1 - DockPanel::DefaultSplitterValue);
			result.Size.x *= DockPanel::DefaultSplitterValue;
			break;
		}
		return result;
	}


	GraphicWindow* DockHintWindow::Proxy::Window = nullptr;
	GraphicWindow* DockHintWindow::Proxy::Left = nullptr;
	GraphicWindow* DockHintWindow::Proxy::Right = nullptr;
	GraphicWindow* DockHintWindow::Proxy::Up = nullptr;
	GraphicWindow* DockHintWindow::Proxy::Down = nullptr;
	GraphicWindow* DockHintWindow::Proxy::Center = nullptr;


	void DockHintWindow::Proxy::InitHitProxy()
	{
		CreateProxy(Left, SE_TEXT("DockHint.Left"));
		CreateProxy(Right, SE_TEXT("DockHint.Right"));
		CreateProxy(Up, SE_TEXT("DockHint.Up"));
		CreateProxy(Down, SE_TEXT("DockHint.Down"));
		CreateProxy(Center, SE_TEXT("DockHint.Center"));
	}
	
	void DockHintWindow::Proxy::Init(Float2& initSize)
	{
		if (Window == nullptr)
		{
			CreateWindowSettings settings = DefaultWindowSettings();
			settings.Title = "DockHint.Window";
			settings.Size = initSize;
			settings.AllowInput = true;
			settings.AllowMaximize = false;
			settings.AllowMinimize = false;
			settings.HasBorder = false;
			settings.HasSizingFrame = false;
			settings.IsRegularWindow = false;
			settings.SupportsTransparency = true;
			settings.ShowInTaskbar = false;
			settings.ShowAfterFirstPaint = false;
			settings.IsTopmost = true;

			Window = New<GraphicWindow>(settings);
			Window->SetOpacity(0.6f);
			Window->GetGUI()->BackgroundColor = Style::Current->DragWindow;
		}
		else
		{
			// Resize proxy
			Window->SetClientSize(initSize);
		}

		InitHitProxy();
	}
	
	void DockHintWindow::Proxy::Hide()
	{
		HideProxy(Window);
		HideProxy(Left);
		HideProxy(Right);
		HideProxy(Up);
		HideProxy(Down);
		HideProxy(Center);
	}
	
	void DockHintWindow::Proxy::Dispose()
	{
		DisposeProxy(Window);
		DisposeProxy(Left);
		DisposeProxy(Right);
		DisposeProxy(Up);
		DisposeProxy(Down);
		DisposeProxy(Center);
	}
	
	void DockHintWindow::Proxy::CreateProxy(GraphicWindow* &win, StringView name)
	{
		if (win != nullptr)
			return;

		CreateWindowSettings settings = DefaultWindowSettings();
		settings.Title = name;
		settings.Size = Float2(HintWindowsSize * Platform::GetDpiScale());
		settings.AllowInput = false;
		settings.AllowMaximize = false;
		settings.AllowMinimize = false;
		settings.HasBorder = false;
		settings.HasSizingFrame = false;
		settings.IsRegularWindow = false;
		settings.SupportsTransparency = true;
		settings.ShowInTaskbar = false;
		settings.ActivateWhenFirstShown = false;
		settings.IsTopmost = true;
		settings.ShowAfterFirstPaint = false;

		win = New<GraphicWindow>(settings);
		win->SetOpacity(0.6f);
		win->GetGUI()->BackgroundColor = Style::Current->DragWindow;
	}
	
	void DockHintWindow::Proxy::HideProxy(GraphicWindow* win)
	{
		if (win)
		{
			win->Hide();
		}
	}

	void DockHintWindow::Proxy::DisposeProxy(GraphicWindow* win)
	{
		if (win)
		{
			win->Close(ClosingReason::User);
			win = nullptr;
		}
	}
} // SE

File: Src/Engine/Editor/GUI/Docking/DockHintWindow.cs
using System;
using SE.GUI;

namespace SE.Editor.GUI
{
    public sealed class DockHintWindow
    {
        private const float BorderMargin = 4.0f;
        private readonly FloatWindowDockPanel m_ToMove;
        private Float2 m_DragOffset;
        private Float2 m_DefaultWindowSize;
        private bool m_LateDragOffsetUpdate;
        private Rectangle m_DockRect;
        private Rectangle m_WindowRect;
        private Rectangle m_UpperRect;
        private Rectangle m_BottomRect;
        private Rectangle m_LeftRect;
        private Rectangle m_RightRect;
        private Rectangle m_CenterRect;

        private DockHintWindow(FloatWindowDockPanel toMove, Float2? initialMouseScreen = null)
        {
            m_ToMove = toMove ?? throw new ArgumentNullException(nameof(toMove));
            TargetState = DockState.Float;
            SE.Window? host = toMove.HostWindow;
            if (host == null)
                throw new InvalidOperationException("Floating dock panel has no host window.");

            m_ToMove.Focus();
            m_ToMove.Defocus();
            host.Focus();
            if (host.IsMaximized)
            {
                Float2 localMouse = host.MousePosition;
                Float2 restoredMouseScreen = host.ClientToScreen(localMouse);
                Float2 previousSize = host.Size;
                host.Restore();
                Float2 restoredSize = host.Size;
                host.Position = new Float2(
                    restoredMouseScreen.X - localMouse.X * restoredSize.X / Math.Max(previousSize.X, 1.0f),
                    restoredMouseScreen.Y - localMouse.Y * restoredSize.Y / Math.Max(previousSize.Y, 1.0f));
            }

            // Floating panels sample after focus, matching the native window mouse query.
            Float2 mouseScreen = initialMouseScreen ?? GetMouseScreenPosition(host);
            if (mouseScreen != Float2.Zero)
                CalculateDragOffset(mouseScreen);
            else
                m_LateDragOffsetUpdate = true;

            // The native implementation keeps the restored size for the whole drag operation.
            m_DefaultWindowSize = host.Size;
            Proxy.Init(m_DefaultWindowSize, host.DpiScale);
            Active = this;
            Proxy.Window?.StartTrackingMouse(false);
            Proxy.Window?.GUI.PerformLayout();
            Update(mouseScreen);
            host.Hide();
            Proxy.Show();
        }

        public static DockHintWindow? Active { get; private set; }
        public FloatWindowDockPanel ToMove => m_ToMove;
        public DockPanel? TargetPanel { get; private set; }
        public DockState TargetState { get; private set; }
        public Rectangle PreviewBounds => m_WindowRect;
        public bool IsDisposed { get; private set; }

        public static DockHintWindow Create(FloatWindowDockPanel toMove)
        {
            if (toMove == null)
                throw new ArgumentNullException(nameof(toMove));
            DockHintWindow hint = new DockHintWindow(toMove);
            return hint;
        }

        public static DockHintWindow Create(DockWindow toMove)
        {
            if (toMove == null)
            {
                throw new ArgumentNullException(nameof(toMove));
            }

            Float2 mouseScreen = toMove.Root is WindowRootControl sourceRoot
                ? GetMouseScreenPosition(sourceRoot.Window)
                : Float2.Zero;
            toMove.ShowFloating();
            FloatWindowDockPanel? floating = toMove.ParentDockPanel as FloatWindowDockPanel;
            if (floating == null)
            {
                throw new InvalidOperationException("The window did not create a floating dock panel.");
            }
            if (floating.HostWindow != null)
            {
                SE.Window host = floating.HostWindow;
                if (mouseScreen == Float2.Zero)
                    mouseScreen = GetMouseScreenPosition(host);
                host.Position = mouseScreen - new Float2(8.0f);
            }
            return new DockHintWindow(floating, mouseScreen);
        }

        public static Rectangle CalculateDockRect(DockState state, Rectangle rect)
        {
            return state switch
            {
                DockState.DockFill => new Rectangle(rect.X, rect.Y + DockPanel.DefaultHeaderHeight,
                    rect.Width, Math.Max(0.0f, rect.Height - DockPanel.DefaultHeaderHeight)),
                DockState.DockTop => new Rectangle(rect.X, rect.Y, rect.Width, rect.Height * DockPanel.DefaultSplitterValue),
                DockState.DockBottom => new Rectangle(rect.X, rect.Y + rect.Height * (1.0f - DockPanel.DefaultSplitterValue),
                    rect.Width, rect.Height * DockPanel.DefaultSplitterValue),
                DockState.DockLeft => new Rectangle(rect.X, rect.Y, rect.Width * DockPanel.DefaultSplitterValue, rect.Height),
                DockState.DockRight => new Rectangle(rect.X + rect.Width * (1.0f - DockPanel.DefaultSplitterValue), rect.Y,
                    rect.Width * DockPanel.DefaultSplitterValue, rect.Height),
                _ => rect,
            };
        }

        public void Update(Float2 mouseScreenPosition)
        {
            if (IsDisposed)
                return;

            TargetPanel = m_ToMove.MasterPanel.HitTest(mouseScreenPosition, m_ToMove);
            TargetState = DockState.Float;
            if (TargetPanel != null)
            {
                m_DockRect = TargetPanel.DockAreaBounds;
                float hintSize = Proxy.HintWindowsSize * (TargetPanel.Root?.DpiScale ?? 1.0f);
                Float2 center = m_DockRect.Location + m_DockRect.Size * 0.5f;
                m_UpperRect = new Rectangle(center.X - hintSize * 0.5f, m_DockRect.Y + BorderMargin, hintSize, hintSize);
                m_BottomRect = new Rectangle(center.X - hintSize * 0.5f, m_DockRect.Bottom - hintSize - BorderMargin, hintSize, hintSize);
                m_LeftRect = new Rectangle(m_DockRect.X + BorderMargin, center.Y - hintSize * 0.5f, hintSize, hintSize);
                m_RightRect = new Rectangle(m_DockRect.Right - hintSize - BorderMargin, center.Y - hintSize * 0.5f, hintSize, hintSize);
                m_CenterRect = new Rectangle(center.X - hintSize * 0.5f, center.Y - hintSize * 0.5f, hintSize, hintSize);

                if (m_ToMove.ChildPanelsCount == 0)
                {
                    if (m_UpperRect.Contains(mouseScreenPosition))
                        TargetState = DockState.DockTop;
                    else if (m_BottomRect.Contains(mouseScreenPosition))
                        TargetState = DockState.DockBottom;
                    else if (m_LeftRect.Contains(mouseScreenPosition))
                        TargetState = DockState.DockLeft;
                    else if (m_RightRect.Contains(mouseScreenPosition))
                        TargetState = DockState.DockRight;
                    if (m_CenterRect.Contains(mouseScreenPosition))
                        TargetState = DockState.DockFill;
                }
            }

            m_WindowRect = TargetState == DockState.Float
                ? new Rectangle(mouseScreenPosition - m_DragOffset, m_DefaultWindowSize)
                : CalculateDockRect(TargetState, m_DockRect);
            bool showHints = TargetPanel != null;
            bool canDockContents = showHints && m_ToMove.ChildPanelsCount == 0;
            Proxy.Update(m_WindowRect, m_UpperRect, m_BottomRect, m_LeftRect, m_RightRect, m_CenterRect,
                showHints, canDockContents, canDockContents);
        }

        public void OnMouseMove(Float2 mouseScreenPosition)
        {
            if (m_LateDragOffsetUpdate)
            {
                CalculateDragOffset(mouseScreenPosition);
                m_LateDragOffsetUpdate = false;
            }
            Update(mouseScreenPosition);
        }
        public void OnMouseUp(Float2 mouseScreenPosition)
        {
            Update(mouseScreenPosition);
            Dispose(true);
        }
        public void OnLostFocus() => Dispose(true);

        public void SetTarget(DockPanel? panel, DockState state)
        {
            TargetPanel = panel;
            TargetState = state;
        }

        public void Dispose() => Dispose(true);

        private void Dispose(bool commit)
        {
            if (IsDisposed)
                return;
            IsDisposed = true;
            Proxy.Window?.EndTrackingMouse();
            Proxy.Hide();

            if (commit && TargetPanel != null && TargetState != DockState.Float)
            {
                bool noChildren = m_ToMove.ChildPanelsCount == 0;
                if (noChildren && m_ToMove.TabsCount == 1)
                {
                    m_ToMove.GetTab(0).Show(TargetState, TargetPanel);
                }
                else if (noChildren && TargetState == DockState.DockFill)
                {
                    while (m_ToMove.TabsCount > 0)
                        m_ToMove.GetTab(0).Show(DockState.DockFill, TargetPanel);
                }
                else
                {
                    DockWindow? selected = m_ToMove.SelectedTab;
                    DockWindow first = m_ToMove.GetTab(0);
                    first.Show(TargetState, TargetPanel);
                    while (m_ToMove.TabsCount > 0)
                        m_ToMove.GetTab(0).Show(DockState.DockFill, first);
                    selected?.SelectTab();
                }
                if (TargetPanel.Root is WindowRootControl root)
                    root.Window.Focus();
            }
            else if (commit && m_ToMove.HostWindow != null)
            {
                Float2 mouseScreen = m_WindowRect.Location + m_DragOffset;
                if (Proxy.Window != null)
                    mouseScreen = Proxy.Window.ClientToScreen(Proxy.Window.GUI.MousePosition * Proxy.Window.DpiScale);
                m_ToMove.HostWindow.Position = mouseScreen - m_DragOffset;
                m_ToMove.HostWindow.Show();
            }

            TargetPanel = null;
            TargetState = DockState.Unknown;
            if (ReferenceEquals(Active, this))
                Active = null;
        }

        private void CalculateDragOffset(Float2 mouseScreenPosition)
        {
            if (m_ToMove.HostWindow != null)
                m_DragOffset = mouseScreenPosition - m_ToMove.HostWindow.Position;
        }

        private static Float2 GetMouseScreenPosition(SE.Window? window)
        {
            if (window == null)
                return Float2.Zero;
            Float2 mouse = window.MousePosition;
            return mouse != Float2.Zero && mouse != Float2.Minimum
                ? window.ClientToScreen(mouse)
                : Float2.Zero;
        }

        public static class Proxy
        {
            private static SE.Window? s_Window;
            private static SE.Window? s_Left;
            private static SE.Window? s_Right;
            private static SE.Window? s_Up;
            private static SE.Window? s_Down;
            private static SE.Window? s_Center;

            public const float HintWindowsSize = 32.0f;
            public static bool IsInitialized { get; private set; }
            public static Float2 Size { get; private set; }
            public static Rectangle PreviewBounds { get; private set; }
            public static SE.Window? Window => s_Window;
            public static SE.Window? Left => s_Left;
            public static SE.Window? Right => s_Right;
            public static SE.Window? Up => s_Up;
            public static SE.Window? Down => s_Down;
            public static SE.Window? Center => s_Center;

            public static void InitHitProxy()
            {
                InitHitProxy(1.0f);
            }

            private static void InitHitProxy(float dpiScale)
            {
                Float2 hintSize = new Float2(HintWindowsSize * dpiScale);
                s_Left ??= CreateProxy("DockHint.Left", false, hintSize);
                s_Right ??= CreateProxy("DockHint.Right", false, hintSize);
                s_Up ??= CreateProxy("DockHint.Up", false, hintSize);
                s_Down ??= CreateProxy("DockHint.Down", false, hintSize);
                s_Center ??= CreateProxy("DockHint.Center", false, hintSize);
                IsInitialized = true;
            }

            public static void Init(Float2 initSize)
            {
                Init(initSize, 1.0f);
            }

            internal static void Init(Float2 initSize, float dpiScale)
            {
                s_Window ??= CreateProxy("DockHint.Window", true, initSize);
                IsInitialized = true;
                Size = initSize;
                PreviewBounds = new Rectangle(Float2.Zero, initSize);
                s_Window.ClientSize = initSize;
                InitHitProxy(dpiScale);
            }

            public static void Show()
            {
                Window?.Show();
                Window?.Focus();
            }

            public static void Update(
                Rectangle preview,
                Rectangle upper,
                Rectangle bottom,
                Rectangle left,
                Rectangle right,
                Rectangle center,
                bool showHints,
                bool showBorders,
                bool showCenter)
            {
                PreviewBounds = preview;
                if (Window != null)
                    Window.ClientBounds = preview;
                SetHint(Up, upper, showHints && showBorders);
                SetHint(Down, bottom, showHints && showBorders);
                SetHint(Left, left, showHints && showBorders);
                SetHint(Right, right, showHints && showBorders);
                SetHint(Center, center, showHints && showCenter);
            }

            public static void Hide()
            {
                PreviewBounds = Rectangle.Empty;
                Window?.Hide();
                Left?.Hide();
                Right?.Hide();
                Up?.Hide();
                Down?.Hide();
                Center?.Hide();
            }

            public static void Dispose()
            {
                CloseProxy(ref s_Window);
                CloseProxy(ref s_Left);
                CloseProxy(ref s_Right);
                CloseProxy(ref s_Up);
                CloseProxy(ref s_Down);
                CloseProxy(ref s_Center);
                IsInitialized = false;
                PreviewBounds = Rectangle.Empty;
                Size = Float2.Zero;
            }

            private static SE.Window CreateProxy(string title, bool allowInput, Float2 size)
            {
                CreateWindowSettings settings = SE.Window.CreateDefaultSettings();
                settings.Title = title;
                settings.Size = size;
                settings.AllowInput = allowInput;
                settings.AllowMaximize = false;
                settings.AllowMinimize = false;
                settings.HasBorder = false;
                settings.HasSizingFrame = false;
                settings.IsRegularWindow = false;
                settings.SupportsTransparency = true;
                settings.ShowInTaskbar = false;
                settings.ActivateWhenFirstShown = allowInput;
                settings.IsTopmost = true;
                settings.ShowAfterFirstPaint = false;

                SE.Window window = SE.Window.Create(settings);
                window.Opacity = 0.6f;
                window.GUI.BackgroundColor = Style.Current.DragWindow;
                if (allowInput)
                {
                    window.MouseMove += OnMouseMove;
                    window.MouseUp += OnMouseUp;
                    window.LostFocus += OnLostFocus;
                }
                return window;
            }

            private static void SetHint(SE.Window? window, Rectangle bounds, bool visible)
            {
                if (window == null)
                    return;
                window.Position = bounds.Location;
                if (visible)
                    window.Show();
                else
                    window.Hide();
            }

            private static void CloseProxy(ref SE.Window? window)
            {
                if (window == null)
                    return;
                window.Close(SE.ClosingReason.User);
                window = null;
            }

            private static void OnMouseMove(ref Float2 localPosition)
            {
                if (Window != null)
                    Active?.OnMouseMove(Window.ClientToScreen(localPosition * Window.DpiScale));
            }

            private static void OnMouseUp(ref Float2 localPosition, MouseButton button, ref bool handled)
            {
                if (button == MouseButton.Left)
                {
                    handled = true;
                    if (Window != null)
                        Active?.OnMouseUp(Window.ClientToScreen(localPosition * Window.DpiScale));
                }
            }

            private static void OnLostFocus()
            {
                Active?.OnLostFocus();
            }
        }
    }
}
